cassiopeia/data/CassiopeiaTree.py

```diff
@@ -2046,6 +2061,564 @@ def get_unmutated_characters_along_edge(
+                        ) / (1 - heritable_proportion)
+                if heritable_missing_rate is None:
+                    heritable_missing_rate = (
+                        -np.log(
+                            (1 - total_missing_proportion)
+                            / (1 - stochastic_missing_rate)
+                        )
+                        / mean_time
+                    )
+                if mutation_rate is None:
+                    mutation_rate = -np.log(1 - mutation_proportion) / mean_time
+
+            if heritable_missing_rate < 0:
+                raise CassiopeiaTreeWarning(
+                    "Estimate of the heritable missing rate using "
+                    "the stochastic rate resulted in a negative "
+                    "heritable missing rate."
+                )
+            if stochastic_missing_rate < 0:
+                raise CassiopeiaTreeWarning(
+                    "Estimate of the stochastic missing rate using "
+                    "the heritable rate resulted in a negative "
+                    "stochastic missing rate."
+                )
+
+        if not use_branch_lengths:
+            mutation_probability_function_of_time = lambda rate, t: rate
+            missing_probability_function_of_time = lambda rate, t: rate
+
+        else:
+            mutation_probability_function_of_time = lambda rate, t: 1 - np.exp(
+                -rate * t
+            )
+            missing_probability_function_of_time = lambda rate, t: 1 - np.exp(
+                -rate * t
+            )
+
+        # We track the likelihoods for each state, at each character,
+        # at each node
+        likelihoods_at_nodes = {}
+
+        # Perform a DFS to propagate the likelihood from the leaves
+        for n in self.depth_first_traverse_nodes(postorder=True):
+            states_at_n = self.get_character_states(n)
+            # If states are observed, their likelihoods are set to 1
+            if self.is_leaf(n):
+                likelihoods_at_nodes[n] = [{i: 1} for i in states_at_n]
+                continue
+
+            possible_states = {}
+            # If internal character states are to be used, then the likelihood
+            # for all other states are ignored. Otherwise, marginalize over
+            # only states that do not break irreversibility, as all states that
+            # do have likelihood of 0
+            if use_internal_character_states:
+                possible_states = [[i] for i in states_at_n]
+            else:
+                for char in range(self.n_character):
+                    child_possible_states = []
+                    for c in [
+                        set(likelihoods_at_nodes[child][char])
+                        for child in self.children(n)
+                    ]:
+                        if (
+                            self.missing_state_indicator not in c
+                            and "*" not in c
```

Comment:
Why do we need the concept of `"*"` in the implementation? It makes things a lot harder to understand for me. Can we avoid it completely?